### PR TITLE
Add navigation sidebar for `dex web`

### LINKED ISSUE #400: Navigation sidebar for Web Notebook
We should have a navigation sidebar for the web notebook, to really leverage the literate programming.

---
image

---

I put together the quick mock-up shown above in javascript based on code from http://loyc.net/2014/javascript-toc.html
Can't use that actual code since it is LGPL, but I basically just ran the first snippet there there and added some stuff to make it float inplace while I scrolled
```javascript
addTOC($get("#main-output"))
sidebox_style = $get(".sidebox").style
sidebox_style.position="fixed"
sidebox_style.right = 0
sidebox_style.top = 0
sidebox_style.marginTop="7rem"
sidebox_style.borderLeft="1px solid firebrick"
sidebox_style.paddingLeft="0.5rem"
sidebox_style.paddingRight="1rem"
```

idk if doing this in javascript is the best way, but seems like it works.
I don't know enough javascript to make a proper version of this if javascript is the way to go.
Could probably work it out though

### PR DESCRIPTION
**Pull request by oxinabox**
closes #400 

As one can probably tell this was:
 1.  written from scratch without reference to someone else implementation
 2. Basically my first time writing javascript
 
It gets the job done, but I am sure it could be way better.
The CSS also could use some work
e.g. right now if it is very long don't know what happens, 
and it will happily go over the top of the main content.

screenshot

### CODE REVIEW

File: static/dynamic.js [hunk 1]
```diff
@@ -77,9 +69,93 @@ source.onmessage = function(event) {
...
+    return structure;
+}
+
+function updateNavigation() {
+    function navItemList(struct) {
+        var listEle = document.createElement('ol')
+        struct.children.forEach(
+            function(childStruct){
+                listEle.appendChild(navItem(childStruct));
+            },
```
Comment (oxinabox) [inline static/dynamic.js:125]: Does Javascriopt have a compose operator?

File: static/dynamic.js [hunk 2]
```diff
@@ -77,9 +69,93 @@ source.onmessage = function(event) {
...
+        );
+        return listEle;
+    }
+    function navItem(struct) {
+        var a = document.createElement('a');  
+        a.appendChild(document.createTextNode(struct["text"]));
```
Comment (oxinabox) [inline static/dynamic.js:131]:
```suggestion
        a.appendChild(document.createTextNode(struct.text));
```

**Comment by dan-zheng** [comment]
Could you please share a screenshot or GIF to the PR description? That would be truly appreciated!

**Comment by oxinabox** [comment]
> Could you please share a screenshot or GIF to the PR description? That would be truly appreciated!

I have added now

I didn't bother since it looks basically like the one in #400 
I wrote the CSS for that and hardly changed it when I made this PR.
I am sure somneone else could write much better CSS

I am going to organize the prelude presently, and then we will see a real example

**Comment by dan-zheng** [comment]
Navigation sidebar for the Prelude sounds useful indeed!

**Comment by oxinabox** [comment]
This is what the prelude would look like. (except for the errors, I don't remember how to run dex without the prelude already loaded)
It doesn't quite all fit on screen so i had to add scrolling 
image

**Comment by dan-zheng** [comment]
That looks really neat - awesome! I imagine it wouldn't too hard to tweak the sidebar's CSS (e.g. indentation width), if we wanted to for some reason?

> (except for the errors, I don't remember how to run dex without the prelude already loaded)

You can try something like:
```
dex --prelude /dev/null web lib/prelude.dx
```

File: static/style.css [hunk 3]
```diff
@@ -13,6 +13,27 @@ body {
...
+  padding-right: 1rem;
+  
+}
+nav:before {
+  content: "Contents";
+  font-weight: bold;
+}
+nav ol {
+  list-style-type:none;
+  padding-left: 1rem;
```
Comment (oxinabox) [inline static/style.css:34]: @dan-zheng this is where you would change indentation width.

My skills go:

- CSS: I am merely bad at, but I know how many things work
- HTML: Making it up as I go along
- Javascript: I literally don't know what semicolons do.

Comment (dan-zheng) [reply static/style.css:34]: Awesome qualifications, thank you for interviewing. When could you start at our $150k full-stack development position?

Comment (dan-zheng) [reply static/style.css:34]: Great, thank you!

**Comment by oxinabox** [comment]
Messed around with the CSS a lot.
Its really fiddly to get the it so one can have a fixed in position sidebar,
while also having horisontally scroll-able main content;
and ensuring it never overlaps.
Probably someone who is better at CSS than me and make that work
But I figure if your screen is narrow enough to need horisonally scrolling, then it is also narrow enough that you don't want to see the sidebar.

### My laptop full screen
fullscreen

### My laptop split screen

split screen


Someone moe competent than me should probably mess with the CSS more

**Comment by dan-zheng** [comment]
https://github.com/google-research/dex-lang/pull/487#pullrequestreview-579308896 renamed some files, like turning `static/dynamic.js` into a unified `static/index.js` exposing a `render(renderMode)` function. Do you have some time to resolve the conflicts?

**Comment by oxinabox** [comment]
> #487 (review) renamed some files, like turning static/dynamic.js into a unified static/index.js exposing a render(renderMode) function. Do you have some time to resolve the conflicts?

Done.
Should (if I have done it right) also add the ToC to the static renderings.
Which will be very useful for tuitorials.

**Comment by oxinabox** [comment]
bump

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks Lyndon!

**Pull request merged**